crates/ruff_python_parser/src/lexer.rs

```diff
@@ -428,6 +429,40 @@ impl<'source> Lexer<'source> {
                     self.cursor.bump();
                     value.push('\\');
                 }
+                '?' => {
+                    self.cursor.bump();
+                    let mut question_count: u8 = 1;
+                    while self.cursor.eat_char('?') {
+                        question_count += 1;
+                    }
+
+                    // The help end magic command is valid if it has 1 or 2 question marks.
+                    // If the `value` is empty i.e., we are at the start of the line,
+                    // then we can't be sure if it is a help end magic command or not.
+                    if question_count > 2
+                        || value.chars().next_back().map_or(true, is_python_whitespace)
+                        || !matches!(self.cursor.first(), '\n' | '\r' | EOF_CHAR)
+                    {
+                        // Not a help end magic command, so continue with the lexing.
+                        value.reserve(question_count as usize);
+                        for _ in 0..question_count {
+                            value.push('?');
+                        }
+                        continue;
+                    }
+
+                    if kind.is_help() {
+                        value = value.trim_start_matches([' ', '?']).to_string();
+                    } else if kind.is_magic() {
+                        value.insert_str(0, &kind.to_string());
```

Comment:
Nit: Can we add a `as_str` method to `MagicKind` to avoid the potential allocation of `to_string` (I don't know if the compiler is able to see through the whole `format!` logic to know that it returns a static `str`)